kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/DataEncryptionKey.java

```diff
@@ -0,0 +1,333 @@
+                    if (LOGGER.isWarnEnabled()) {
+                        LOGGED_DESTROY_FAILED.computeIfAbsent(cls, (c) -> {
+                            LOGGER.warn("Failed to destroy an instance of {}. "
+                                    + "Note: this message is logged once per class even though there may be many occurrences of this event. "
+                                    + "This event can happen because the JRE's SecretKeySpec class does not override the destroy() method.",
+                                    c, e);
+                            return Boolean.TRUE;
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    public boolean isDestroyed() {
+        SecretKey secretKey = atomicKey.get();
+        return secretKey == null || secretKey.isDestroyed();
+    }
+
+    /**
+     * A means of performing a limited number of encryption operations without access to key material.
+     */
+    @NotThreadSafe
+    public final class Encryptor implements AutoCloseable {
+        private final Cipher cipher;
+        private SecretKey key;
+        private final Supplier<AlgorithmParameterSpec> paramSupplier;
+        private final CipherSpec cipherSpec;
+        private int numEncryptions;
+
+        private Encryptor(CipherSpec cipherSpec, SecretKey key, int numEncryptions) {
+            if (numEncryptions <= 0) {
+                throw new IllegalArgumentException();
+            }
+            this.cipher = cipherSpec.newCipher();
+            this.paramSupplier = cipherSpec.paramSupplier();
+            this.cipherSpec = cipherSpec;
+            this.key = key;
+            this.numEncryptions = numEncryptions;
+        }
+
+        /**
+         * @return The encrypted key
+         */
+        public E edek() {
+            return edek;
+        }
+
+        /**
+         * Perform an encryption operation using the DEK.
+         * @param plaintext The plaintext to be encrypted
+         * @param aad The AAD to be included in the encryption
+         * @param paramAllocator A function that will return a buffer into which the cipher parameters will be written.
+         * The function's argument is the number of bytes required for the cipher parameters.
+         * @param ciphertextAllocator A function that will return a buffer into which the ciphertext will be written.
+         * The function's first argument is the number of bytes required for the cipher parameters.
+         * The function's second argument is the number of bytes required for the ciphertext.
+         * @throws DekUsageException If this Encryptor has run out of operations.
+         */
+        public void encrypt(@NonNull ByteBuffer plaintext,
+                            @Nullable ByteBuffer aad,
+                            @NonNull IntFunction<ByteBuffer> paramAllocator,
+                            @NonNull BiFunction<Integer, Integer, ByteBuffer> ciphertextAllocator) {
```

Comment:
* I didn't love the API we have here, but it works for our purposes. 
* Perhaps these should both be `BiFunction`s (i.e. compute both sizes before calling the allocators)?
* I also don't love that we're using `BiFunction`: Its contract is much looser than what we require here, so perhaps declaring our own interfaces and documenting its contract would be better?

Reply:
+1 to replacing bifunctions with our own functional interfaces

Reply:
I did this one